dotCMS/src/main/java/com/dotcms/rest/api/v1/site/SiteResource.java

```diff
@@ -244,4 +297,633 @@ public final Response switchSite(
         }
 
     }
-} // E:O:F:SiteBrowserResource.
\ No newline at end of file
+
+    /**
+     * Retrieve the host thumbnails
+     * @param httpServletRequest
+     * @param httpServletResponse
+     * @return
+     * @throws PortalException
+     * @throws SystemException
+     * @throws DotDataException
+     * @throws DotSecurityException
+     */
+    @GET
+    @Path("/thumbnails")
+    @JSONP
+    @NoCache
+    @Produces({MediaType.APPLICATION_JSON, "application/javascript"})
+    public Response findAllSiteThumbnails(@Context final HttpServletRequest httpServletRequest,
+                                @Context final HttpServletResponse httpServletResponse) throws PortalException, SystemException, DotDataException, DotSecurityException {
+
+        final User user = new WebResource.InitBuilder(this.webResource)
+                .requestAndResponse(httpServletRequest, httpServletResponse)
+                .requiredBackendUser(true)
+                .rejectWhenNoUser(true)
+                .init().getUser();
+
+        final ContentletAPI contentletAPI = APILocator.getContentletAPI();
+        final PageMode      pageMode      = PageMode.get(httpServletRequest);
+        final boolean respectFrontend     = pageMode.respectAnonPerms;
+        final List<Host> hosts            = this.siteHelper.findAll(user, respectFrontend);
+        final HostNameComparator hostNameComparator = new HostNameComparator();
+
+        Logger.debug(this, ()-> "Finding all site thumbnails...");
+
+        return Response.ok(new ResponseEntityView(hosts.stream().filter(DotLambdas.not(Host::isSystemHost))
+                .sorted(hostNameComparator).map(host -> this.toSiteMap(user, contentletAPI, host))
+                .collect(Collectors.toList()))).build();
+    }
+
+    private Map<String, Object> toSiteMap(final User user,
+                                          final ContentletAPI contentletAPI, final Host host) {
+
+        final Map<String, Object> thumbInfo = new HashMap<>();
+        thumbInfo.put("hostId",    host.getIdentifier());
+        thumbInfo.put("hostInode", host.getInode());
+        thumbInfo.put("hostName",  host.getHostname());
+        final File hostThumbnail   = Try.of(()-> contentletAPI.getBinaryFile(host.getInode(), Host.HOST_THUMB_KEY, user)).getOrNull();
+        final boolean hasThumbnail = hostThumbnail != null;
+        thumbInfo.put("hasThumbnail", hasThumbnail);
+        thumbInfo.put("tagStorage",   host.getMap().get("tagStorage"));
+        return thumbInfo;
+    }
+
+    /**
+     * Publish a host
+     * @param httpServletRequest
+     * @param httpServletResponse
+     * @param hostId
+     * @return
+     * @throws DotDataException
+     * @throws DotSecurityException
+     * @throws PortalException
+     * @throws SystemException
+     */
+    @PUT
+    @Path("/_publish/{hostId}")
```

Comment:
I think it's better to have the hostId and then the action the exec. `/{hostId}/_publish`

Reply:
Yeah. A bit of a pain to change them but I agree it's the proper way.